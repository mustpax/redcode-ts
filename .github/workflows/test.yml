name: Tests

on:
  push:
    branches: [main]
  pull_request:
    branches: [main]

jobs:
  test:
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v4

      - name: Setup Bun
        uses: oven-sh/setup-bun@v1
        with:
          bun-version: latest

      - name: Install dependencies
        run: bun install

      - name: Build parser from grammar
        run: bun run build

      - name: Run tests
        run: bun test
