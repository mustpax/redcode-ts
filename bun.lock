{
  "lockfileVersion": 1,
  "workspaces": {
    "": {
      "name": "redcode-ts",
      "dependencies": {
        "peggy": "^4.2.0",
      },
      "devDependencies": {
        "@types/bun": "latest",
      },
      "peerDependencies": {
        "typescript": "^5.0.0",
      },
    },
  },
  "packages": {
    "@peggyjs/from-mem": ["@peggyjs/from-mem@1.3.5", "", { "dependencies": { "semver": "7.6.3" } }, "sha512-oRyzXE7nirAn+5yYjCdWQHg3EG2XXcYRoYNOK8Quqnmm+9FyK/2YWVunwudlYl++M3xY+gIAdf0vAYS+p0nKfQ=="],

    "@types/bun": ["@types/bun@1.2.4", "", { "dependencies": { "bun-types": "1.2.4" } }, "sha512-QtuV5OMR8/rdKJs213iwXDpfVvnskPXY/S0ZiFbsTjQZycuqPbMW8Gf/XhLfwE5njW8sxI2WjISURXPlHypMFA=="],

    "@types/node": ["@types/node@22.13.9", "", { "dependencies": { "undici-types": "~6.20.0" } }, "sha512-acBjXdRJ3A6Pb3tqnw9HZmyR3Fiol3aGxRCK1x3d+6CDAMjl7I649wpSd+yNURCjbOUGu9tqtLKnTGxmK6CyGw=="],

    "@types/ws": ["@types/ws@8.5.14", "", { "dependencies": { "@types/node": "*" } }, "sha512-bd/YFLW+URhBzMXurx7lWByOu+xzU9+kb3RboOteXYDfW+tr+JZa99OyNmPINEGB/ahzKrEuc8rcv4gnpJmxTw=="],

    "bun-types": ["bun-types@1.2.4", "", { "dependencies": { "@types/node": "*", "@types/ws": "~8.5.10" } }, "sha512-nDPymR207ZZEoWD4AavvEaa/KZe/qlrbMSchqpQwovPZCKc7pwMoENjEtHgMKaAjJhy+x6vfqSBA1QU3bJgs0Q=="],

    "commander": ["commander@12.1.0", "", {}, "sha512-Vw8qHK3bZM9y/P10u3Vib8o/DdkvA2OtPtZvD871QKjy74Wj1WSKFILMPRPSdUSx5RFK1arlJzEtA4PkFgnbuA=="],

    "peggy": ["peggy@4.2.0", "", { "dependencies": { "@peggyjs/from-mem": "1.3.5", "commander": "^12.1.0", "source-map-generator": "0.8.0" }, "bin": { "peggy": "bin/peggy.js" } }, "sha512-ZjzyJYY8NqW8JOZr2PbS/J0UH/hnfGALxSDsBUVQg5Y/I+ZaPuGeBJ7EclUX2RvWjhlsi4pnuL1C/K/3u+cDeg=="],

    "semver": ["semver@7.6.3", "", { "bin": { "semver": "bin/semver.js" } }, "sha512-oVekP1cKtI+CTDvHWYFUcMtsK/00wmAEfyqKfNdARm8u1wNVhSgaX7A8d4UuIlUI5e84iEwOhs7ZPYRmzU9U6A=="],

    "source-map-generator": ["source-map-generator@0.8.0", "", {}, "sha512-psgxdGMwl5MZM9S3FWee4EgsEaIjahYV5AzGnwUvPhWeITz/j6rKpysQHlQ4USdxvINlb8lKfWGIXwfkrgtqkA=="],

    "typescript": ["typescript@5.8.2", "", { "bin": { "tsc": "bin/tsc", "tsserver": "bin/tsserver" } }, "sha512-aJn6wq13/afZp/jT9QZmwEjDqqvSGp1VT5GVg+f/t6/oVyrgXM6BY1h9BRh/O5p3PlUPAe+WuiEZOmb/49RqoQ=="],

    "undici-types": ["undici-types@6.20.0", "", {}, "sha512-Ny6QZ2Nju20vw1SRHe3d9jVu6gJ+4e3+MMpqu7pqE5HT6WsTSlce++GQmK5UXS8mzV8DSYHrQH+Xrf2jVcuKNg=="],
  }
}
