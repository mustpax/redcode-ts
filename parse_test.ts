import { test, expect } from "bun:test";

test("snap", () => {
  expect("foo").toMatchSnapshot();
});
