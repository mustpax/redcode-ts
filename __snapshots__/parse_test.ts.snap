// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`parse all examples empty.redcode 1`] = `
[
  {
    "comment": null,
    "instruction": null,
  },
]
`;

exports[`parse all examples comment.redcode 1`] = `
[
  {
    "comment": " just comment",
    "instruction": null,
  },
]
`;

exports[`parse all examples dat.redcode 1`] = `
[
  {
    "comment": null,
    "instruction": {
      "args": null,
      "modifier": null,
      "opcode": "DAT",
    },
  },
  {
    "comment": null,
    "instruction": {
      "args": [
        {
          "mmode": null,
          "value": 1,
        },
        {
          "mmode": null,
          "value": 1,
        },
      ],
      "modifier": null,
      "opcode": "JMP",
    },
  },
]
`;

exports[`parse all examples long.redcode 1`] = `
[
  {
    "comment": null,
    "instruction": {
      "args": [
        {
          "mmode": null,
          "value": 0,
        },
        {
          "mmode": null,
          "value": 0,
        },
      ],
      "modifier": null,
      "opcode": "DAT",
    },
  },
  {
    "comment": " this inst kills current process",
    "instruction": {
      "args": null,
      "modifier": null,
      "opcode": "DAT",
    },
  },
  {
    "comment": null,
    "instruction": {
      "args": [
        {
          "mmode": "$",
          "value": 1,
        },
        {
          "mmode": "}",
          "value": 0,
        },
      ],
      "modifier": "X",
      "opcode": "MOV",
    },
  },
  {
    "comment": " comment text",
    "instruction": null,
  },
  {
    "comment": null,
    "instruction": {
      "args": null,
      "modifier": null,
      "opcode": "MOV",
    },
  },
  {
    "comment": null,
    "instruction": null,
  },
  {
    "comment": null,
    "instruction": {
      "args": [
        {
          "mmode": "#",
          "value": 4,
        },
        {
          "mmode": null,
          "value": 3,
        },
      ],
      "modifier": null,
      "opcode": "ADD",
    },
  },
  {
    "comment": null,
    "instruction": {
      "args": [
        {
          "mmode": null,
          "value": 2,
        },
        {
          "mmode": "@",
          "value": 2,
        },
      ],
      "modifier": null,
      "opcode": "MOV",
    },
  },
  {
    "comment": null,
    "instruction": {
      "args": [
        {
          "mmode": null,
          "value": -2,
        },
        null,
      ],
      "modifier": null,
      "opcode": "JMP",
    },
  },
  {
    "comment": null,
    "instruction": {
      "args": [
        {
          "mmode": "#",
          "value": 0,
        },
        {
          "mmode": "#",
          "value": 0,
        },
      ],
      "modifier": null,
      "opcode": "DAT",
    },
  },
]
`;

exports[`parse all examples empty.redcode - snapshot 1`] = `
[
  {
    "comment": null,
    "instruction": null,
  },
]
`;

exports[`parse all examples comment.redcode - snapshot 1`] = `
[
  {
    "comment": " just comment",
    "instruction": null,
  },
]
`;

exports[`parse all examples dat.redcode - snapshot 1`] = `
[
  {
    "comment": null,
    "instruction": {
      "args": null,
      "modifier": null,
      "opcode": "DAT",
    },
  },
  {
    "comment": null,
    "instruction": {
      "args": [
        {
          "mmode": null,
          "value": 1,
        },
        {
          "mmode": null,
          "value": 1,
        },
      ],
      "modifier": null,
      "opcode": "JMP",
    },
  },
]
`;

exports[`parse all examples long.redcode - snapshot 1`] = `
[
  {
    "comment": null,
    "instruction": {
      "args": [
        {
          "mmode": null,
          "value": 0,
        },
        {
          "mmode": null,
          "value": 0,
        },
      ],
      "modifier": null,
      "opcode": "DAT",
    },
  },
  {
    "comment": " this inst kills current process",
    "instruction": {
      "args": null,
      "modifier": null,
      "opcode": "DAT",
    },
  },
  {
    "comment": null,
    "instruction": {
      "args": [
        {
          "mmode": "$",
          "value": 1,
        },
        {
          "mmode": "}",
          "value": 0,
        },
      ],
      "modifier": "X",
      "opcode": "MOV",
    },
  },
  {
    "comment": " comment text",
    "instruction": null,
  },
  {
    "comment": null,
    "instruction": {
      "args": null,
      "modifier": null,
      "opcode": "MOV",
    },
  },
  {
    "comment": null,
    "instruction": null,
  },
  {
    "comment": null,
    "instruction": {
      "args": [
        {
          "mmode": "#",
          "value": 4,
        },
        {
          "mmode": null,
          "value": 3,
        },
      ],
      "modifier": null,
      "opcode": "ADD",
    },
  },
  {
    "comment": null,
    "instruction": {
      "args": [
        {
          "mmode": null,
          "value": 2,
        },
        {
          "mmode": "@",
          "value": 2,
        },
      ],
      "modifier": null,
      "opcode": "MOV",
    },
  },
  {
    "comment": null,
    "instruction": {
      "args": [
        {
          "mmode": null,
          "value": -2,
        },
        null,
      ],
      "modifier": null,
      "opcode": "JMP",
    },
  },
  {
    "comment": null,
    "instruction": {
      "args": [
        {
          "mmode": "#",
          "value": 0,
        },
        {
          "mmode": "#",
          "value": 0,
        },
      ],
      "modifier": null,
      "opcode": "DAT",
    },
  },
]
`;
